### PR TITLE
EBAX-1047 Add merchant client

### PR DESCRIPTION
**Pull request by junnae**
New client for integrating with merchant api

### CODE REVIEW

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessTokenSubscriber.java [hunk 1]
```diff
@@ -1,6 +1,6 @@
 package no.bankaxept.epayment.client.base.accesstoken;
 
-import no.bankaxept.epayment.client.base.AccessFailed;
+import no.bankaxept.epayment.client.base.exception.AccessFailed;
```
[the remark this answers is not in the document]
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessTokenSubscriber.java:3]: Or actually, I don't know how to handle this.. It's only thrown from the access token subscriber, but we really don't want to expose that package. Do we just use a more generic runtime exception then? I move this up for now..

File: java/merchant/src/main/java/no/bankaxept/epayment/client/merchant/MerchantClient.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    public MerchantClient(BaseClient baseClient) {
+        this.baseClient = baseClient;
+    }
+
+    public MerchantClient(String baseurl, String apimKey, String username, String password) {
+        this.baseClient = new BaseClient(baseurl, apimKey, username, password);
+    }
+
+    public Flow.Publisher<HttpResponse> payment(PaymentRequest request, String correlationId) throws JsonProcessingException {
```
[the remark this answers is not in the document]
Comment (junnae) [reply java/merchant/src/main/java/no/bankaxept/epayment/client/merchant/MerchantClient.java]: I added a new class with mapping now!

File: java/merchant/src/main/java/no/bankaxept/epayment/client/merchant/MerchantClient.java [hunk 3]
```diff
@@ -40,8 +33,10 @@ public MerchantClient(String baseurl, String apimKey, String username, String pa
         this.baseClient = new BaseClient(baseurl, apimKey, username, password);
     }
 
-    public Flow.Publisher<HttpResponse> payment(PaymentRequest request, String correlationId) throws JsonProcessingException {
-        return baseClient.post(PAYMENTS_URL, new SinglePublisher<>(objectMapper.writeValueAsString(request), executor), correlationId);
+    public Flow.Publisher<Void> payment(PaymentRequest request, String correlationId) throws JsonProcessingException {
```
[the remark this answers is not in the document]
Comment (junnae) [reply java/merchant/src/main/java/no/bankaxept/epayment/client/merchant/MerchantClient.java]: I've more or less like you've suggested now. Made a few small adjustments

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java [hunk 4]
```diff
@@ -1,20 +1,38 @@
...
+    public Response toResponse() {
+        switch (status) {
+            case 201:
+                return Response.Accepted;
+            case 422:
+                return Response.Rejected;
+            case 409:
+                return Response.Conflicted;
+            default:
+                if (is2xxOk()) return Response.Repeated;
```
Comment (thxmasj) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java:29]: Too wide. Only 200 indicates repeat. 2xx which is not 200 nor 201 should map to Failed.

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java [hunk 5]
```diff
@@ -1,20 +1,38 @@
...
+                if (is4xxClientError()) return Response.ClientError;
+                if (is5xxServerError()) return Response.Failed;
+                else return Response.Unknown;
```
Comment (thxmasj) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java:32]: I don't think it makes sense to differ between Failed and Unknown. That would only raise questions from the integrator, and our response would be to handle Unknown same way as Failed.

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/AbstractClient.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+        this.baseClient = baseClient;
+    }
+
+    public AbstractClient(String baseurl, String apimKey, String username, String password) {
+        this.baseClient = new BaseClient(baseurl, apimKey, username, password);
+    }
+
+    protected Flow.Publisher<Response> postEmptyResponseBody(String url, SinglePublisher<String> bodyPublisher, String correlationId){
+        return new EmptyResponseProcessor(baseClient.post(url, bodyPublisher, correlationId));
+    }
```
Comment (thxmasj) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/AbstractClient.java:19]: Misnamed. No empty response anymore. And are you sure the AbstractClient class and EmptyResponseProcessor class bring any value? They are kind of just double wrapping the MapOperator without any additional substance, as I see it.

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessTokenProcessor.java [hunk 7]
```diff
@@ -65,6 +65,9 @@ public void onNext(HttpResponse item) {
         AccessToken token;
         try {
             token = AccessToken.parse(item.getBody());
+            if(token.getExpiry().isBefore(clock.instant())) {
+                throw new IllegalStateException("Fetched an already expired token");
+            }
```
[the remark this answers is not in the document]
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessTokenProcessor.java:70]: Can't you just move this check to the test code, then? What you write is not related to the production code at all?

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/Response.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package no.bankaxept.epayment.client.base;
+
+public enum Response {
```
Comment (thxmasj) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/Response.java:3]: I think `RequestStatus` would be a better name. The typed responses come as callbacks.

File: java/merchant-client/src/main/java/no/bankaxept/epayment/client/merchant/MerchantClient.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+    public MerchantClient(BaseClient baseClient) {
+        super(baseClient);
+    }
+
+    public MerchantClient(String baseurl, String apimKey, String username, String password) {
+        super(baseurl, apimKey, username, password);
+    }
+
+    public Flow.Publisher<Response> payment(PaymentRequest request, String correlationId) throws JsonProcessingException {
```
Comment (thxmasj) [inline java/merchant-client/src/main/java/no/bankaxept/epayment/client/merchant/MerchantClient.java:34]: It is cleaner to use an unchecked exception which is not declared (could be RuntimeException wrapping the JsonProcessingException). Then we are free to change the underlying implementation and still be backwards compatible.

**Review by thxmasj — CHANGES_REQUESTED** [review]

File: java/merchant/src/main/java/module-info.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+module no.bankaxept.epayment.sdk.merchant {
+    requires com.fasterxml.jackson.core;
+    requires com.fasterxml.jackson.databind;
+    requires com.fasterxml.jackson.datatype.jsr310;
+    requires no.bankaxept.epayment.sdk.baseclient;
+    requires swagger.annotations;
+    requires java.compiler;
+    requires jsr305;
```
[the remark this answers is not in the document]
Comment (thxmasj) [reply java/merchant/src/main/java/module-info.java]: Maintainers could also provide a name in the manifest, according to [this](https://stackoverflow.com/questions/46501047/what-does-required-filename-based-automodules-detected-warning-mean). Anyhow, we leave it as is.

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessTokenSubscriber.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessTokenSubscriber.java]: We should probably map it to some sensible value for the API methods' response values. Let's leave that for now.

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/Response.java (hunk 8 above)
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/Response.java:3]: It still represents a response though..? I still think `request` is misleading

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java (hunk 4 above)
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java]: changed

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java (hunk 5 above)
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java]: Sure, changed

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/Response.java (hunk 8 above)
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/Response.java]: I called it ResponseStatus now..

File: java/merchant-client/src/main/java/no/bankaxept/epayment/client/merchant/MerchantClient.java (hunk 9 above)
Comment (junnae) [reply java/merchant-client/src/main/java/no/bankaxept/epayment/client/merchant/MerchantClient.java]: Changed this

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessTokenProcessor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessTokenProcessor.java]: I'll just remove it. But I don't think it's impossible to ever have an unexpected value or format in production,

**Review by thxmasj — APPROVED** [review]

**Pull request merged**